src/Plutus/Types/DataSchema.purs

```diff
@@ -0,0 +1,262 @@
+   If a constructor does not have a Record argument, we do not need to encode any information about that argument. For example, the type:
+
+    data GType
+      = G0 BigInt
+      | G1 Boolean Boolean Boolean
+      | G2 BigInt GType
+
+   would have a much simpler Schema:
+
+   instance
+     HasPlutusSchema GType
+       ( "G0" := PNil @@ Z
+       :+ "G1" := PNil @@ (S Z)
+       :+ "G2" :+ PNil @@ (S (S Z))
+       :+ PNil)
+
+   The sole purpose of the inner list is to ensure correct translation to and from Plutus Data for record types
+   when using the generic functions in ToData/FromData. Since GType's constructors do not have record arguments,
+   no additional information is needed in the inner lists.
+-}
+-- | See the source (DataSchema.purs) for a detailed explanation of how this works / what it is for / examples
+type PlutusSchema = PSchema (PSchema Type)
+
+-- | A class used to associate types with a Plutus Data Schema. The fundeps *should* guarantee
+-- | that only one schema can exist for each type. This allows us to make the compiler select the
+-- | schema associated with a particular type.
+class HasPlutusSchema
+  :: Type -> PlutusSchema -> Constraint
+class HasPlutusSchema t schema | t -> schema
+
+-- | Listlike constructors for the PSchema kind.
+foreign import data PNil :: forall (k :: Type). PSchema k
+foreign import data PCons
+  :: forall (k :: Type). Field k -> PSchema k -> PSchema k
+
+-- | Type synonym which is used to provide syntatic sugar to PCons. You can think of :+ as a type level version of : for PSchema
+type ApPCons :: forall (k :: Type). Field k -> PSchema k -> PSchema k
+type ApPCons x xs = PCons x xs
+
+infixr 0 type ApPCons as :+
+
+-- | Indexed kind. A kind representing a type @k@ indexed by a type level natural number @n@.
+data IxK :: forall (k :: Type). k -> Type
+data IxK k
+
+-- | A data type of Kind (IxK k n) for some k and n
+foreign import data MkIxK :: forall (k :: Type). Nat -> k -> IxK k
+
+-- | Syntactic sugar which allows us to represent MkIxK as a type operator
+type MkIxK_
+  :: forall k1. k1 -> Nat -> IxK k1 -- FIXME(@gnumonik): s/k1/k it complains
+type MkIxK_ k n = MkIxK n k
+
+infixr 9 type MkIxK_ as @@
+
+-- | A kind which is used to represent pairs of Symbols and IxKs. A fancy type level tuple, more or less.
+data Field :: forall (k :: Type). k -> Type
+data Field k
+
+-- | A data type of Kind (Field k)
+foreign import data MkField
+  :: forall (k :: Type) (f :: Type -> Type). Symbol -> f k -> Field k
+
+-- | Type synonym used to represent MkField as a type operator.
+type MkField_
+  :: forall (f :: Type -> Type) (k :: Type). Symbol -> f k -> Field k
+type MkField_ lbl ixty = MkField lbl ixty
+
+infixr 8 type MkField_ as :=
+
+{- <<< Constraints used to provide compile-time validation of a PSchema >>> -}
+
+-- TODO: See if we can optimize this. Slows down compilation considerably for complex types.
```

Comment:
We should add an issue for this and put the URL here

Reply:
Done (also added references to the issue to the other Uniqueness classes)